nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java

```diff
@@ -0,0 +1,1076 @@
+                                        failUnmappedColumns, warningUnmappedColumns, escapeColumnNames, quoteTableName);
+                            } else {
+                                log.error("Statement Type {} is not valid, flowfile {} will be penalized and routed to failure", new Object[]{statementType, flowFile});
+                                flowFile = session.penalize(flowFile);
+                                session.transfer(flowFile, REL_FAILURE);
+                                return;
+                            }
+                        } catch (final ProcessException pe) {
+                            log.error("Failed to convert {} to a SQL {} statement due to {}; routing to failure",
+                                    new Object[]{flowFile, statementType, pe.toString()}, pe);
+                            flowFile = session.penalize(flowFile);
+                            session.transfer(flowFile, REL_FAILURE);
+                            return;
+                        }
+
+                        try (PreparedStatement ps = con.prepareStatement(sqlHolder.getSql())) {
+
+                            try {
+                                ps.setQueryTimeout(queryTimeout); // timeout in seconds
+                            } catch (SQLException se) {
+                                // If the driver doesn't support query timeout, then assume it is "infinite". Allow a timeout of zero only
+                                if (queryTimeout > 0) {
+                                    throw se;
+                                }
+                            }
+
+                            Record currentRecord;
+                            List<Integer> fieldIndexes = sqlHolder.getFieldIndexes();
+
+                            while ((currentRecord = recordParser.nextRecord()) != null) {
+                                Object[] values = currentRecord.getValues();
+                                if (values != null) {
+                                    if (fieldIndexes != null) {
+                                        for (int i = 0; i < fieldIndexes.size(); i++) {
+                                            ps.setObject(i + 1, values[fieldIndexes.get(i)]);
+                                        }
+                                    } else {
+                                        // If there's no index map, assume all values are included and set them in order
+                                        for (int i = 0; i < values.length; i++) {
+                                            ps.setObject(i + 1, values[i]);
+                                        }
+                                    }
+                                    ps.addBatch();
+                                }
+                            }
+
+                            log.debug("Executing query {}", new Object[]{sqlHolder});
+                            ps.executeBatch();
```

Comment:
I assumed that this processor can be used not only for CDC use-cases, but also for simply ingesting multiple records as a single batch insertion. It works as expected for normal cases. However when I tested with a FlowFile containing multiple records mixed with successful and failure records, a BatchUpdateException was thrown, and the incoming FlowFile was routed to "retry". Database table has records without the failed ones.

This would cause data duplication if the same FlowFile is processed again, as it inserts the successful records again (it's possible depends on primary/unique key setting).

I put nifi template and SQLs in this [Gist](https://gist.github.com/ijokarumawak/d2ae2d582472780cc5769c19dd204033) to reproduce this behavior.

I would expect this processor to rollback without commit records partially. Or create FlowFiles for each destination relationship containing associated records. How do you think?